lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb

```diff
@@ -1689,6 +1689,10 @@ fc_extras
         if np.issubdtype(cf_var.dtype, np.str_):
             attr_units = cf_units._NO_UNIT_STRING
 
+        if "flag_values" in cf_var._nc_attrs or "flag_masks" in cf_var._nc_attrs \
```

Comment:
I don't think we should be using a private property of CFVariable here if we can avoid it.
I've just spent much too long [trying to understand exactly what CFVariable does + how](https://github.com/SciTools/iris/pull/3725), 
and I would say, please let's not reply on that implementation here !

I think I would use `any(hasattr(cf_var, name) for name in (...))`.

It's true, this will be less efficient because [hasattr depends entirely on getattr](https://stackoverflow.com/a/24971134)
On the other hand, it will only fetch them once per variable, as the CFVariable implementation does ensure that.
Given that netcdf files don't have the proliferation of small things that field-based formats do, I think it will be ok for performance -- although we could test that ?

Reply:
So it looks like there is some unusual behaviour associated with getting atrributes from `cf_var` where accessing an attribute will cause it to not be picked up as an attribute in the final object. I think this is due to this piece of code https://github.com/SciTools/iris/blob/715d01921f01903132538ea2574653155fa4412d/lib/iris/fileformats/cf.py#L175-L183

It seems like in the current implementation, "attributes" contains everything which has not already been accessed in this way. Trying to access these attributes prematurely causes them to be ignored later on.
I believe this is why I chose to take attributes from `cf_var._nc_attrs` rather than directly from `cf_var`. I tested your suggestion and it looks like this would cause attributes to load incorrectly.

This does mean that we are bypassing a mechanism which is specifically designed to improve performance. I have a feeling that the ideal solution may have to involve separating the mechanism for accessing `cf_var` attributes from the mechanism for chosing which `cf_var` attributes to ignore when setting iris attributes.

Reply:
> @stephenworsley I have a feeling that the ideal solution may have to involve separating the mechanism for accessing cf_var attributes from the mechanism for chosing which cf_var attributes to ignore when setting iris attributes.

You're dead right ! 
I do think that the overlapping schemes in CFVariable generate a whole heap of extra confusion.
( I spent quite some time trying to work out why the 'ignored' and 'unused' ideas work in just the the way that they do.  )

Reply:
Having said this, checking the attributes of `cf_var.cf_data` seems to work while relying on more public API.

Reply:
So, I'm really sorry for making a duff suggestion !
A simple "clean" solution would be to test `hasattr(cf_var.nc_variable)` instead of `hasattr(cf_var)`.  But, like you, I'm still a bit concerned about bypassing the speedup mechanism (even though I suspect that at least some of it is probably unnecessary).

Another really simply fix for my original objection would be to retain your existing code solution, but just make `cf_var._nc_attrs` *public* -- by renaming, removing the underscore.  
**How would you feel about that ?**

( Note: there aren't many existing uses of this, and they are all within cf.py )

Reply:
I feel like either of these could work. To be honest, I don't have a strong enough picture of how `cf_var` is structured to have a clear idea of which feels more correct. In either case, I suspect both approaches would require changing the a lot of the tests currently using a mocked `cf_var`.